### PR TITLE
Add convergence criteria to linear solver

### PR DESCRIPTION
**Pull request by nilsvu**
## Proposed changes

Convergence criteria for the linear solver implemented so far:
- Max. number of iterations
- Absolute residual
- Residual relative to the initial residual

Also adds a compute item for `LinearSolver::Tags::HasConverged` that uses the convergence criteria.

Please see the commit messages for more details.

### Types of changes:

- [x] New feature

### Component:

- [x] Code

### Code review checklist

- [ ] The PR passes all checks, including unit tests, `clang-tidy` and `IWYU`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### CODE REVIEW

File: src/NumericalAlgorithms/LinearSolver/Tags.hpp [hunk 1]
```diff
@@ -75,7 +83,7 @@ struct HasConverged : db::SimpleTag {
   static std::string name() noexcept {
     return "LinearSolverHasConverged";
   }
-  using type = bool;
+  using type = std::pair<bool, std::string>;
```
Comment (nilsvu) [inline src/NumericalAlgorithms/LinearSolver/Tags.hpp]: Note to self: update docs for this Tag

File: src/NumericalAlgorithms/LinearSolver/ConvergenceCriteria.hpp [hunk 2]
```diff
@@ -0,0 +1,110 @@
...
+    static constexpr OptionString help = {
+        "The residual has reached this magnitude."};
+    static type lower_bound() noexcept { return 0.; }
+    static type default_value() noexcept { return 1e-14; }
+  };
+
+  struct RelativeResidual {
+    using type = double;
+    static constexpr OptionString help = {
+        "The residual has decreased by this factor."};
```
Comment (hrueter) [inline src/NumericalAlgorithms/LinearSolver/ConvergenceCriteria.hpp]: The option descriptions in this function are confusing.

E.g. here I think you mean something like
"The solver terminates when the residual decreased by this factor."

The past tense in your text would imply, that this condition is somehow already met at the start of the iteration.

**Comment by wthrowe** [comment]
I haven't looked at the code in detail, but have you considered doing this using the events-and-triggers infrastructure?

**Comment by nilsvu** [comment]
@wthrowe I have not, indeed I thought the events-and-triggers are not quite ready for use yet. For this particular purpose (linear solver convergence criteria) they are probably not applicable, except perhaps to replace the `TerminateIfConverged` action. That's because the convergence happens on a singleton parallel component (the residual monitor) in a reduction action, and not on the elements in an algorithm action.

**Comment by wthrowe** [comment]
Events and triggers have been ready for use since last February, but they were mostly intended for use in an action iteration sequence so you may be right that they can't be used here.

File: src/NumericalAlgorithms/LinearSolver/ConvergenceCriteria.hpp (hunk 2 above)
Comment (nilsvu) [reply src/NumericalAlgorithms/LinearSolver/ConvergenceCriteria.hpp]: The option strings must be very short, so I chose this style of documentation to complement the option string of the struct. I'm not convinced to change it yet 😄

**Comment by nilsvu** [comment]
I split the changes into more descriptive commits for easier review

**Comment by kidder** [comment]
@nilsleiffischer this fails to compile

**Comment by nilsvu** [comment]
Made some more changes, in particular added a struct that represents the convergence status instead of a `std::pair<bool, std::string>`.

This is now ready for review. @nilsdeppe we discussed this so perhaps you could review, or forward to someone else? @hrueter would you also have time to review?

**Comment by nilsdeppe** [comment]
@nilsleiffischer sure, but it still doesn't compile on Travis. I'm going to wait for it to be compiling and passing tests before review. Marking as "in progress" for now until that's fixed :)

File: src/NumericalAlgorithms/LinearSolver/Convergence.hpp [hunk 3]
```diff
@@ -0,0 +1,169 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+/// \file
+/// Defines DataBox tags for the linear solver
```
Comment (wthrowe) [inline src/NumericalAlgorithms/LinearSolver/Convergence.hpp]: Remove.

File: src/NumericalAlgorithms/LinearSolver/Convergence.hpp [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+
+/*!
+ * \brief Criteria that determine the linear solve has converged
+ *
+ * \details The following criteria are implemented:
+ *
+ * - MaxIterations: Matches if the number of iterations exceeds this limit.
+ * - AbsoluteResidual: Matches if the residual has reached this magnitude.
+ * Bounded below by the product of the machine epsilon and the conditioning
+ * number of the linear operator that is being inverted.
```
Comment (wthrowe) [inline src/NumericalAlgorithms/LinearSolver/Convergence.hpp]: I don't understand what the second sentence ("Bounded below...") is trying to tell me.

File: src/NumericalAlgorithms/LinearSolver/Convergence.hpp [hunk 5]
```diff
@@ -0,0 +1,169 @@
...
+class er;
+}  // namespace PUP
+/// \endcond
+
+namespace LinearSolver {
+
+/*!
+ * \brief Criteria that determine the linear solve has converged
+ *
+ * \details The following criteria are implemented:
```
Comment (wthrowe) [inline src/NumericalAlgorithms/LinearSolver/Convergence.hpp]: Do you want to specify which one "wins" if multiple criteria are satisfied on the same iteration?

File: tests/Unit/NumericalAlgorithms/LinearSolver/Test_Convergence.cpp [hunk 6]
```diff
@@ -0,0 +1,123 @@
...
+              criteria, LinearSolver::IterationId{1}, 0., 1.) ==
+          LinearSolver::ConvergenceReason::AbsoluteResidual);
+    CHECK(LinearSolver::convergence_criteria_match(
+              criteria, LinearSolver::IterationId{1}, 1., 2.) ==
+          LinearSolver::ConvergenceReason::RelativeResidual);
+  }
+
+  {
+    INFO("HasNotConverged");
+    CHECK_FALSE(LinearSolver::HasConverged{});
```
Comment (wthrowe) [inline tests/Unit/NumericalAlgorithms/LinearSolver/Test_Convergence.cpp]: Test serialization of a default-constructed object.

File: tests/InputFiles/Poisson/Input2D.yaml [hunk 7]
```diff
@@ -12,14 +12,19 @@ DomainCreator:
     LowerBound: [0, 0]
     UpperBound: [3.141592653589793, 3.141592653589793]
     IsPeriodicIn: [false, false]
-    InitialRefinement: [1, 1]
+    InitialRefinement: [0, 0]
```
Comment (wthrowe) [inline tests/InputFiles/Poisson/Input2D.yaml]: Was this change intentional?  It seems unrelated to the rest of the commit.

File: src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitorActions.hpp [hunk 8]
```diff
@@ -58,51 +54,32 @@ struct InitializeResidualMagnitude {
...
-    // When more sophisticated convergence criteria are implemented we may also
-    // want to check for convergence here.
+    // Determine whether the linear solver has already converged. This invokes
+    // the compute item.
+    const auto& has_converged = db::get<LinearSolver::Tags::HasConverged>(box);
+
+    if (UNLIKELY(has_converged and
+                 static_cast<int>(get<::Tags::Verbosity>(box)) >=
+                     static_cast<int>(::Verbosity::Quiet))) {
+      Parallel::printf(MakeString{} << has_converged);
```
Comment (wthrowe) [inline src/NumericalAlgorithms/LinearSolver/Gmres/ResidualMonitorActions.hpp]: `Parallel::printf("%s", has_converged)`  There are several other similar bits of code in this PR, which should also be changed.

The important part of this change is the `"%s"`.  The fact that you can (I think) avoid the `MakeString` is a side benefit.  If the string form of `has_converged` happens to contain a "%" then the current code may segfault.  The first argument to `printf` should (almost) always be a string literal.  (This is GCC's `-Wformat-security` warning for the C `printf`.)

**Review by wthrowe — CHANGES_REQUESTED** [review]

File: src/NumericalAlgorithms/LinearSolver/Convergence.hpp (hunk 4 above)
Comment (nilsvu) [reply src/NumericalAlgorithms/LinearSolver/Convergence.hpp]: I added more documentation. Does this make it more clear?

**Comment by nilsvu** [comment]
Addressed your comments @wthrowe. @nilsdeppe or @hrueter could you review?

File: src/NumericalAlgorithms/LinearSolver/Convergence.hpp [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+ *
+ * The following criteria are implemented, ordered from highest to lowest
+ * priority:
+ *
+ * - AbsoluteResidual: Matches if the residual has reached this magnitude.
+ * - RelativeResidual: Matches if the residual has decreased by this factor,
+ * relative to the start of the first linear solver iteration.
+ * - MaxIterations: Matches if the number of iterations exceeds this limit.
+ *
+ * \note that the smallest possible residual magnitude the linear solver can
```
Comment (wthrowe) [inline src/NumericalAlgorithms/LinearSolver/Convergence.hpp]: Start a new sentence after each `\note`.  The rendered documentation does not format in a way that "Note" can be read as part of the sentence.  (It is completely omitted in some cases, even.)

**Review by wthrowe — CHANGES_REQUESTED** [review]
The new documentation is much clearer.  Aside from the minor formatting comment, I am happy with this.

**Comment by nilsvu** [comment]
Thanks for reviewing @wthrowe, squashed and rebased.

**Comment by nilsvu** [comment]
ping @nilsdeppe (or forward please)

File: src/NumericalAlgorithms/LinearSolver/Convergence.hpp [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+ * relative to the start of the first linear solver iteration.
+ * - MaxIterations: Matches if the number of iterations exceeds this limit.
+ *
+ * \note The smallest possible residual magnitude the linear solver can reach is
+ * the product between the machine epsilon and the condition number of the
+ * linear operator that is being inverted. Smaller residuals are numerical
+ * artifacts. Requiring an absolute or relative residual below this limit will
+ * likely lead to termination by `MaxIterations`.
+ *
+ * \note Remembere that when the linear operator \f$A\f$ corresponds to a PDE
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/LinearSolver/Convergence.hpp]: `Remembere` -> `Remember`

File: src/NumericalAlgorithms/LinearSolver/Convergence.cpp [hunk 11]
```diff
@@ -0,0 +1,149 @@
...
+    : reason_(convergence_criteria_match(convergence_criteria, iteration_id,
+                                         residual_magnitude,
+                                         initial_residual_magnitude)),
+      convergence_criteria_(convergence_criteria),
+      iteration_id_(iteration_id),
+      residual_magnitude_(residual_magnitude),
+      initial_residual_magnitude_(initial_residual_magnitude) {}
+
+ConvergenceReason HasConverged::reason() const noexcept {
+  ASSERT(reason_,
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/LinearSolver/Convergence.cpp:85]: I think this should be an `ERROR` because the `*reason_` will throw and I've noticed it's usually quite difficult to figure out what happened when exception is thrown but not caught. I also don't expect that we will pay much in CPU time for changing this to an ERROR and I can see it making somebody's life easier :)

File: tests/Unit/NumericalAlgorithms/LinearSolver/Test_Tags.cpp [hunk 12]
```diff
@@ -32,4 +40,38 @@ SPECTRE_TEST_CASE("Unit.Numerical.LinearSolver.Tags",
         "LinearOrthogonalizationHistory(Tag)");
   CHECK(LinearSolver::Tags::KrylovSubspaceBasis<Tag>::name() ==
         "KrylovSubspaceBasis(Tag)");
+  CHECK(LinearSolver::Tags::ConvergenceCriteria::name() ==
+        "ConvergenceCriteria");
+
+  {
+    INFO("HasConvergedCompute");
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/LinearSolver/Test_Tags.cpp:49]: Can you add a check for `::name()` for this and the compute tag as well please? Thanks!

File: tests/Unit/NumericalAlgorithms/LinearSolver/Test_Tags.cpp [hunk 13]
```diff
@@ -74,4 +76,13 @@ SPECTRE_TEST_CASE("Unit.Numerical.LinearSolver.Tags",
     CHECK(db::get<LinearSolver::Tags::HasConverged>(box).reason() ==
           LinearSolver::ConvergenceReason::AbsoluteResidual);
   }
+
+  {
+    INFO("MagnitudeCompute");
+    const auto box =
```
Comment (nilsdeppe) [inline tests/Unit/NumericalAlgorithms/LinearSolver/Test_Tags.cpp:86]: Could you please add a test of `::name()`?

File: src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp [hunk 14]
```diff
@@ -35,53 +35,65 @@ namespace cg_detail {
...
     db::mutate<residual_square_tag>(
-        make_not_null(&box), [res_new](const gsl::not_null<double*>
-                                           res_old) noexcept {
-          *res_old = res_new;
+        make_not_null(&box), [residual_square](
+                                 const gsl::not_null<double*>
+                                     local_residual_square) noexcept {
+          *local_residual_square = residual_square;
         });
+    // Perform a separate `db::mutate` so that we can retrieve the
```
Comment (nilsdeppe) [inline src/NumericalAlgorithms/LinearSolver/ConjugateGradient/ResidualMonitorActions.hpp:67]: :+1: on this comment :)

**Review by nilsdeppe — CHANGES_REQUESTED** [review]
A few minor things, I'm happy for you to squash immediately :)

File: src/NumericalAlgorithms/LinearSolver/Convergence.cpp (hunk 11 above)
Comment (nilsvu) [reply src/NumericalAlgorithms/LinearSolver/Convergence.cpp:85]: I'm not sure if I understand correctly. The `ASSERT` tests if the `boost::optional<ConvergenceReason> reason_` evaluates to `true`, in which case we are sure that unpacking it with `*` will succeed (and never throw).
I `ASSERT` this because I make clear in the documentation that it is a programming error to call the `reason()` function without first making sure that `HasConverged::bool()` is indeed true.

File: tests/Unit/NumericalAlgorithms/LinearSolver/Test_Tags.cpp (hunk 12 above)
Comment (nilsvu) [reply tests/Unit/NumericalAlgorithms/LinearSolver/Test_Tags.cpp:49]: The `::name()` of the simple tag `HasConverged` is tested above already. The compute tag derives from it, so should I test its `::name()` too?

**Comment by nilsvu** [comment]
should be done @nilsdeppe @wthrowe

File: src/NumericalAlgorithms/LinearSolver/Convergence.cpp (hunk 11 above)
Comment (nilsdeppe) [reply src/NumericalAlgorithms/LinearSolver/Convergence.cpp:85]: okay

File: tests/Unit/NumericalAlgorithms/LinearSolver/Test_Tags.cpp (hunk 12 above)
Comment (nilsdeppe) [reply tests/Unit/NumericalAlgorithms/LinearSolver/Test_Tags.cpp:49]: Whatever you prefer

**Comment by nilsdeppe** [comment]
Looks like you might have a fun rebase :)

**Comment by nilsvu** [comment]
It was indeed fun ;)

While rebasing on #1269 I added another commit that fixes some observing (last commit). Could you have a look at that one @nilsdeppe and @wthrowe?

**Review by nilsdeppe — DISMISSED** [review]

**Review by wthrowe — DISMISSED** [review]

**Comment by nilsdeppe** [comment]
@nilsleiffischer IWYU is unhappy

**Comment by nilsvu** [comment]
@nilsdeppe @wthrowe fixed iwyu. should be done.

**Review by nilsdeppe — APPROVED** [review]

**Review by wthrowe — APPROVED** [review]

**Comment by nilsvu** [comment]
@sxs-collaboration/spectre-core-devs this is ready

**Comment by nilsdeppe** [comment]
Travis failures are the usual GCC debug segfaults.
